libsplinter/src/collections/ref_map.rs

```diff
@@ -54,21 +55,20 @@ impl RefMap {
     /// Decrements the referece count for `ref_id`
     ///
     /// If the internal reference count reaches zero, then `ref_id` will be removed.
-    pub fn remove_ref(&mut self, ref_id: &str) -> Result<Option<String>, RefMapRemoveError> {
+    pub fn remove_ref(&mut self, ref_id: &K) -> Result<Option<K>, RefMapRemoveError> {
         let ref_count = match self.references.remove(ref_id) {
             Some(ref_count) => ref_count,
             None => {
-                return Err(RefMapRemoveError(format!(
-                    "Trying to remove a reference that does not exist: {}",
-                    ref_id
-                )))
+                return Err(RefMapRemoveError(
+                    "Trying to remove a reference that does not exist".into(),
+                ))
             }
         };
 
         if ref_count == 1 {
-            Ok(Some(ref_id.into()))
+            Ok(Some(ref_id.clone()))
```

Comment:
These `clone` and `to_string` calls might be avoidable if you follow a similar pattern to how structures like the standard library's HashMap.  See [HashMap::get](https://doc.rust-lang.org/std/collections/struct.HashMap.html#method.get) as an example.

Reply:
Updated. Clone is now no longer required and a str can be passed to remove_ref now. Because of how the remove_ref returns the K (not the value) and updates the value if not 0, I needed to switch to using `remove_entry` instead of `remove` to have access to the key.